jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChinesePostman.java

```diff
@@ -0,0 +1,302 @@
+     *
+     * @param graph input graph
+     * @return Eulerian Circuit
+     */
+    private GraphPath<V,E> solveCPPUndirected(Graph<V, E> graph){
+
+        //1. Find all odd degree vertices (there should be an odd number of those)
+        List<V> oddDegreeVertices=graph.vertexSet().stream().filter(v ->graph.degreeOf(v)%2==1).collect(Collectors.toList());
+
+        //2. Compute all pairwise shortest paths for the oddDegreeVertices
+        Map<Pair<V,V>, GraphPath<V,E>> shortestPaths=new HashMap<>();
+        ShortestPathAlgorithm<V,E> sp=new DijkstraShortestPath<>(graph);
+        for(int i=0; i<oddDegreeVertices.size()-1; i++){
+            for(int j=i+1; j<oddDegreeVertices.size(); j++){
+                V u=oddDegreeVertices.get(i);
+                V v=oddDegreeVertices.get(j);
+                shortestPaths.put(new UnorderedPair<>(u, v), sp.getPath(u, v));
+            }
+        }
+
+        //3. Solve a matching problem. For that we create an auxiliary graph.
+        Graph<V, DefaultWeightedEdge> auxGraph=new SimpleWeightedGraph<>(DefaultWeightedEdge.class);
+        Graphs.addAllVertices(auxGraph, oddDegreeVertices);
+
+        for(V u : oddDegreeVertices){
+            for(V v : oddDegreeVertices){
+                if(u==v) continue;
+                Graphs.addEdge(auxGraph, u, v, shortestPaths.get(new UnorderedPair<>(u,v)).getWeight());
+            }
+        }
+        MatchingAlgorithm.Matching<V, DefaultWeightedEdge> matching =new KolmogorovMinimumWeightPerfectMatching<>(auxGraph).getMatching();
+
+        //4. On the original graph, add shortcuts between the odd vertices. These shortcuts have been
+        //identified by the matching algorithm. A shortcut from u to v
+        //indirectly implies duplicating all edges on the shortest path from u to v
+        Graph<V,E> eulerGraph=new Pseudograph<>(graph.getVertexSupplier(), graph.getEdgeSupplier(), graph.getType().isWeighted());
+        Graphs.addAllVertices(eulerGraph, graph.vertexSet()); Graphs.addAllEdges(eulerGraph, graph, graph.edgeSet());
+        Map<E, GraphPath<V,E>> shortcutEdges=new HashMap<>();
+        for(DefaultWeightedEdge e: matching.getEdges()){
+            V u=auxGraph.getEdgeSource(e);
+            V v=auxGraph.getEdgeTarget(e);
+            E shortcutEdge =eulerGraph.addEdge(u, v);
+            shortcutEdges.put(shortcutEdge, shortestPaths.get(new UnorderedPair<>(u,v)));
+        }
+
+        EulerianCycleAlgorithm<V, E> eulerianCycleAlgorithm=new HierholzerEulerianCycle<>();
+        GraphPath<V,E> pathWithShortcuts=eulerianCycleAlgorithm.getEulerianCycle(eulerGraph);
+        return replaceShortcutEdges(graph, pathWithShortcuts, shortcutEdges);
+    }
+
+
+    /**
+     * Solves the CPP for directed graphs
+     * 
+     * @param graph input graph
+     * @return Eulerian Circuit
+     */
+    private GraphPath<V, E> solveCPPDirected(Graph<V, E> graph)
+    {
+
+        // 1. Find all imbalanced vertices (there should be an odd number of those)
```

Comment:
(even)

Reply:
deleted. There's not necessary an odd number of imbalanced vertices.